File: wip/srv/check_top_client.py
import threading

import rclpy
from rclpy.node import Node
from rclpy.task import Future
from std_srvs.srv import Trigger


class CheckTopClient(Node):
  def __init__(self):
    super().__init__("check_top_client")
    self.cli = self.create_client(Trigger, "/is_ball_on_top")

    while not self.cli.wait_for_service(timeout_sec=0.5):
      self.get_logger().info("service not available, waiting again...")

    self.future: Future = None

  def send_request(self):
    if self.future is not None and not self.future.done():
      self.future.cancel()  # Cancel the future. The callback will be called with Future.result == None.

      self.get_logger().info(
        "Service Future canceled. The Node took too long to process the service call."
      )

    req = Trigger.Request()
    self.future = self.cli.call_async(req)
    self.future.add_done_callback(self.process_response)

    # Start a timer for the timeout
    self.timeout_thread = threading.Timer(1.0, self.check_timeout)
    self.timeout_thread.start()

  def process_response(self, future: Future):
    if self.timeout_thread.is_alive():
      self.timeout_thread.cancel()

    response: Trigger.Response = future.result()
    if response is None:
      self.get_logger().info("Response is None")
      return
    self.get_logger().info(f"Success: {response.success}")
    self.get_logger().info(f"Response: {response.message}")

  def check_timeout(self):
    if self.future and not self.future.done():
      self.future.cancel()
      self.get_logger().info("Service call timed out.")


def main(args=None):
  try:
    rclpy.init(args=args)

    check_top_client = CheckTopClient()

    rclpy.spin(check_top_client)

  except KeyboardInterrupt:
    pass

  except Exception as e:
    print(e)


if __name__ == "__main__":
  main()

File: wip/srv/check_top_client.cpp
#include <chrono>
#include <memory>
#include <rclcpp/rclcpp.hpp>
#include <std_srvs/srv/trigger.hpp>

using namespace std::chrono_literals;

class CheckTopClient : public rclcpp::Node
{
public:
  CheckTopClient()
      : Node("check_top_client")
  {
    client_ = this->create_client<std_srvs::srv::Trigger>("/is_ball_on_top");

    while (!client_->wait_for_service(500ms))
    {
      RCLCPP_INFO(this->get_logger(), "Service not available, waiting again...");
    }
  }

  void send_request()
  {
    if (future_ && !future_->is_ready())
    {
      RCLCPP_INFO(this->get_logger(), "Previous service call timed out or was not completed. Canceling...");
      future_.reset();
      timeout_timer_->cancel();
    }

    auto request = std::make_shared<std_srvs::srv::Trigger::Request>();
    future_ = client_->async_send_request(request, std::bind(&CheckTopClient::process_response, this, std::placeholders::_1));

    timeout_timer_ = this->create_wall_timer(1s, std::bind(&CheckTopClient::check_timeout, this));
  }

private:
  void process_response(rclcpp::Client<std_srvs::srv::Trigger>::SharedFuture future)
  {
    if (timeout_timer_ && timeout_timer_->is_canceled() == false)
    {
      timeout_timer_->cancel();
    }

    auto response = future.get();
    if (response)
    {
      RCLCPP_INFO(this->get_logger(), "Success: %d", response->success);
      RCLCPP_INFO(this->get_logger(), "Response: %s", response->message.c_str());
    }
    else
    {
      RCLCPP_INFO(this->get_logger(), "Service call was unsuccessful.");
    }
  }

  void check_timeout()
  {
    if (future_ && !future_->is_ready())
    {
      RCLCPP_INFO(this->get_logger(), "Service call timed out.");
      future_.reset();
    }
  }

  rclcpp::Client<std_srvs::srv::Trigger>::SharedPtr client_;
  rclcpp::Client<std_srvs::srv::Trigger>::SharedFuture future_;
  rclcpp::TimerBase::SharedPtr timeout_timer_;
};

int main(int argc, char **argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<CheckTopClient>();

  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
